File: tests/unit/test_transition_cache.py
import json

import numpy as np
import pytest

from engram.transitions.cache import CacheFormatError, TransitionCache


def _put(cache, state, token, *, source="online", next_state=None):
    state = np.asarray(state, dtype=np.float32)
    if next_state is None:
        next_state = state + 1.0
    return cache.put(
        state,
        token,
        next_state,
        [(token + 1, 2.5), (token + 2, 1.0)],
        0.9,
        source=source,
    )


def test_fingerprint_is_deterministic_and_input_token_is_part_of_key():
    cache = TransitionCache(
        4, quantization_step=0.5, subvector_width=2, similarity_radius=0.5
    )
    state = np.array([0.1, -0.1, 1.1, -1.1])
    assert cache.fingerprint(state, 7) == cache.fingerprint(state.copy(), 7)
    assert cache.fingerprint(state, 7) != cache.fingerprint(state, 8)

    _put(cache, state, 7)
    nearby = state + np.array([0.02, 0.0, 0.0, 0.0])
    lookup = cache.lookup(nearby, 7)
    assert lookup.hit
    assert lookup.transition.output_candidates[0] == (8, 2.5)
    assert not cache.lookup(nearby, 8).hit


def test_fingerprint_collision_outside_radius_is_rejected():
    cache = TransitionCache(
        2, quantization_step=10.0, similarity_radius=0.1, subvector_width=1
    )
    _put(cache, [1.0, 0.0], 3)
    rejected = cache.lookup([2.0, 0.0], 3)

    assert not rejected.hit
    assert rejected.reason == "radius_rejection"
    assert rejected.state_distance == pytest.approx(0.5)
    assert cache.metrics.radius_rejections == 1
    assert cache.metrics.collisions == 1


def test_lru_eviction_online_offline_counts_and_error_metrics():
    cache = TransitionCache(2, capacity=2, quantization_step=0.1, similarity_radius=0.01)
    _put(cache, [1.0, 0.0], 1, source="offline")
    _put(cache, [0.0, 1.0], 2, source="online")
    assert cache.lookup([1.0, 0.0], 1).hit  # Refresh token 1 as most recent.
    _put(cache, [-1.0, 0.0], 3, source="online")

    assert not cache.lookup([0.0, 1.0], 2).hit
    measured = cache.lookup([1.0, 0.0], 1, actual_next_state=[2.1, 1.0])
    assert measured.hit
    metrics = cache.metrics
    assert metrics.entries == 2
    assert metrics.evictions == 1
    assert metrics.offline_puts == 1
    assert metrics.online_puts == 2
    assert metrics.approximation_error_samples == 1
    assert metrics.mean_approximation_error > 0.0
    assert metrics.hit_rate == pytest.approx(2.0 / 3.0)


def test_bypass_mode_suppresses_population_and_reuse():
    cache = TransitionCache(2, bypass=True)
    assert not _put(cache, [1.0, 0.0], 4)
    lookup = cache.lookup([1.0, 0.0], 4)
    assert not lookup.hit and lookup.reason == "bypass"
    assert cache.metrics.entries == 0
    assert cache.metrics.bypassed_puts == 1
    assert cache.metrics.bypassed_lookups == 1


def test_persistent_round_trip_preserves_lru_data_and_detects_corruption(tmp_path):
    cache = TransitionCache(
        3,
        capacity=3,
        quantization_step=0.25,
        subvector_width=2,
        similarity_radius=0.1,
    )
    _put(cache, [1.0, 0.0, 0.0], 10, source="offline")
    _put(cache, [0.0, 1.0, 0.0], 11, source="online")
    path = cache.save(tmp_path / "cache.json")

    restored = TransitionCache.load(path)
    result = restored.lookup([1.0, 0.0, 0.0], 10)
    assert result.hit
    np.testing.assert_array_equal(result.transition.next_state, [2.0, 1.0, 1.0])
    assert result.transition.confidence == 0.9
    assert restored.metrics.offline_puts == 0  # Loading does not count as population.

    document = json.loads(path.read_text())
    document["entries"][0]["confidence"] = 0.1
    path.write_text(json.dumps(document))
    with pytest.raises(CacheFormatError, match="checksum mismatch"):
        TransitionCache.load(path)

File: native/tests/test_transition_cache.cpp
#include "engram/transition_cache.h"

#include <array>
#include <cmath>
#include <iostream>
#include <stdexcept>

namespace {

constexpr std::array<engram::TransitionCandidate, 2> kCandidates = {
    engram::TransitionCandidate{7, 2.5F},
    engram::TransitionCandidate{8, 1.0F}};

int fail(const char* message) {
  std::cerr << message << '\n';
  return 1;
}

bool close(const double left, const double right) {
  return std::abs(left - right) < 1e-6;
}

}  // namespace

int main() {
  {
    engram::TransitionCache cache(4, 4, 1.0F, 1.0F, 2);
    const std::array<float, 4> state = {0.5F, 1.5F, 2.5F, -1.5F};
    const auto fingerprint = cache.fingerprint(state, 9);
    const std::array<std::int16_t, 4> expected = {0, 2, 2, -2};
    if (fingerprint.input_token != 9 ||
        !std::equal(fingerprint.codes.begin(), fingerprint.codes.end(),
                    expected.begin()) ||
        fingerprint == cache.fingerprint(state, 10)) {
      return fail("deterministic ties-to-even fingerprint mismatch");
    }
  }

  {
    engram::TransitionCache cache(2, 4, 0.5F, 0.5F, 2);
    const std::array<float, 2> state = {0.1F, -0.1F};
    const std::array<float, 2> next = {1.0F, 2.0F};
    if (!cache.put_offline(state, 3, next, kCandidates, 0.9F)) {
      return fail("offline cache population failed");
    }
    const std::array<float, 2> nearby = {0.11F, -0.1F};
    const auto hit = cache.lookup(nearby, 3);
    if (!hit.hit || hit.reason != engram::TransitionLookupReason::kHit ||
        hit.transition.output_candidates[0].token != 7 ||
        hit.transition.next_state[1] != 2.0F ||
        cache.lookup(nearby, 4).hit) {
      return fail("validated transition lookup mismatch");
    }
  }

  {
    engram::TransitionCache cache(2, 2, 10.0F, 0.1F, 2);
    const std::array<float, 2> stored = {1.0F, 0.0F};
    const std::array<float, 2> query = {2.0F, 0.0F};
    const std::array<float, 2> next = {0.0F, 1.0F};
    cache.put_online(stored, 5, next, kCandidates, 1.0F);
    const auto rejected = cache.lookup(query, 5);
    const auto metrics = cache.metrics();
    if (rejected.hit ||
        rejected.reason != engram::TransitionLookupReason::kRadiusRejection ||
        !close(rejected.state_distance, 0.5) ||
        metrics.radius_rejections != 1 || metrics.collisions != 1) {
      return fail("fingerprint collision radius rejection mismatch");
    }
  }

  {
    engram::TransitionCache cache(2, 2, 0.1F, 0.01F, 2);
    const std::array<float, 2> a = {1.0F, 0.0F};
    const std::array<float, 2> b = {0.0F, 1.0F};
    const std::array<float, 2> c = {-1.0F, 0.0F};
    const std::array<float, 2> next_a = {2.0F, 1.0F};
    const std::array<float, 2> next_b = {1.0F, 2.0F};
    const std::array<float, 2> next_c = {0.0F, 1.0F};
    cache.put_offline(a, 1, next_a, kCandidates, 0.8F);
    cache.put_online(b, 2, next_b, kCandidates, 0.8F);
    static_cast<void>(cache.lookup(a, 1));  // Refresh A.
    cache.put_online(c, 3, next_c, kCandidates, 0.8F);
    if (cache.lookup(b, 2).hit) {
      return fail("LRU did not evict the oldest entry");
    }
    const std::array<float, 2> actual = {2.1F, 1.0F};
    if (!cache.lookup(a, 1, actual).hit) {
      return fail("LRU evicted refreshed entry");
    }
    const auto metrics = cache.metrics();
    if (metrics.entries != 2 || metrics.evictions != 1 ||
        metrics.offline_puts != 1 || metrics.online_puts != 2 ||
        metrics.approximation_error_samples != 1 ||
        metrics.mean_approximation_error <= 0.0 ||
        !close(metrics.hit_rate, 2.0 / 3.0)) {
      return fail("LRU or transition metrics mismatch");
    }
  }

  {
    engram::TransitionCache cache(2, 2, 0.1F, 0.1F, 2, true);
    const std::array<float, 2> state = {1.0F, 0.0F};
    const std::array<float, 2> next = {0.0F, 1.0F};
    if (cache.put_online(state, 1, next, kCandidates, 1.0F)) {
      return fail("bypass accepted cache population");
    }
    const auto result = cache.lookup(state, 1);
    const auto metrics = cache.metrics();
    if (result.hit || result.reason != engram::TransitionLookupReason::kBypass ||
        metrics.entries != 0 || metrics.bypassed_puts != 1 ||
        metrics.bypassed_lookups != 1) {
      return fail("transition bypass metrics mismatch");
    }
  }

  try {
    engram::TransitionCache cache(2, 2, 0.1F, 0.1F, 1);
    const std::array<float, 2> state = {1.0F, 0.0F};
    cache.put_online(state, 1, state, kCandidates, 1.0F);
    return fail("oversized output candidate list was accepted");
  } catch (const std::invalid_argument&) {
  }

  return 0;
}
